### PR TITLE
OPDS based languages and categories

### LINKED ISSUE #934: Don't use a static list of language/categories in the catalog filtering.
When we have added the language selector, the catalog hadn't the language list feature.
So we've set a fixed list of languages in kiwix-desktop.

When next version of libkiwix will be released, catalog will have a API to get the list of languages and categories. We should use it.

### PR DESCRIPTION
**Pull request by juuz0**
Fixes #934 
Requires libkiwix from: https://github.com/kiwix/libkiwix/pull/967

### CODE REVIEW

**Review by mgautierfr — CHANGES_REQUESTED** [review]

File: src/contentmanager.cpp [hunk 1]
```diff
@@ -47,6 +54,42 @@ QStringList ContentManager::getTranslations(const QStringList &keys)
...
+        categories.push_back("all");
+        for (auto category : categoryData) {
+            auto categoryName = QString::fromStdString(category.get("name")->string_value());
+            categories.push_back(categoryName);
+        }
+        m_categories = categories;
+        emit(categoriesLoaded(m_categories));
+        return;
+    }
+    m_remoteLibraryManager.getCategoriesFromOpds();
```
[the remark this answers is not in the document]
Comment (juuz0) [reply src/contentmanager.cpp]: Was hesitant with doing changes in libkiwix but done now!

**Comment by kelson42** [comment]
@juuz0 Can you please rebase and update about the status? Are you just missing a new review?

**Comment by juuz0** [comment]
@kelson42 Yes, it's missing a new review. The libkiwix side is merged and done.
@mgautierfr Please!

File: src/contentmanager.cpp [hunk 2]
```diff
@@ -585,6 +587,18 @@ void ContentManager::updateRemoteLibrary(const QString& content) {
     });
 }
 
+void ContentManager::updateLanguages(const QString& content) {
+    m_remoteLibrary = kiwix::Library();
```
Comment (mgautierfr) [inline src/contentmanager.cpp:632]: This re initialize the`m_remoteLibrary`. Why we need this ?
If we update categories after languages, we will lost the known languages.

File: src/opdsrequestmanager.cpp [hunk 3]
```diff
@@ -49,19 +48,62 @@ void OpdsRequestManager::doUpdate(const QString& currentLanguage, const QString&
     qInfo() << "Downloading" << url.toString(QUrl::FullyEncoded);
 
     QNetworkRequest request(url);
-    if (mp_reply) {
-        mp_reply->abort();
-    }
-    mp_reply = m_networkManager.get(request);
-    connect(mp_reply, &QNetworkReply::finished, this, &OpdsRequestManager::receiveContent);
+    auto mp_reply = m_networkManager.get(request);
```
Comment (mgautierfr) [inline src/opdsrequestmanager.cpp:50]: Could we set the builded url (before this line) with `opdsResponseFromPath` ?

File: src/contentmanager.cpp [hunk 4]
```diff
@@ -70,7 +70,7 @@ ContentManager::ContentManager(Library* library, kiwix::Downloader* downloader,
     connect(treeView, &QTreeView::doubleClicked, this, &ContentManager::openBookWithIndex);
     connect(&m_remoteLibraryManager, &OpdsRequestManager::languagesReceived, this, &ContentManager::updateLanguages);
     connect(&m_remoteLibraryManager, &OpdsRequestManager::categoriesReceived, this, &ContentManager::updateCategories);
-    setLanguages();
+    setCategories();
```
Comment (mgautierfr) [inline src/contentmanager.cpp:72]: This comment is for the commit message, not this line.

Commit message say : `If local library is displayed, we get the categories from LibraryDumper`
It is false as the categories are taken from Library itself (and it is a good thing)

**Review by mgautierfr — CHANGES_REQUESTED** [review]

File: src/contentmanager.cpp (hunk 2 above)
Comment (juuz0) [reply src/contentmanager.cpp:632]: This is what happens when I forget things to remove while editing commits in a rebase.
Thanks and fixed.

File: src/opdsrequestmanager.cpp (hunk 3 above)
Comment (juuz0) [reply src/opdsrequestmanager.cpp:50]: Im divided on whats the better way here
1. Have 2 functions:
`opdsResponseFromPath(path)` (which calls the next one with an empty query)
`opdsResponseFromPath(path, query)`

2. Have a default empty query parameter
`opdsResponseFromPath(path, query = "")`

File: src/kiwixapp.h [hunk 5]
```diff
@@ -86,6 +86,7 @@ class KiwixApp : public QtSingleApplication
     QString getText(const QString &key) { return m_translation.getText(key); };
     void setMonitorDir(const QString &dir);
     bool isCurrentArticleBookmarked();
+    kiwix::UpdatableNameMapper* getNameMapper() { return &m_nameMapper; }
```
[the remark this answers is not in the document]
Comment (juuz0) [reply src/kiwixapp.h:89]: None. Forgot to remove while changing methods

File: src/contentmanager.cpp (hunk 4 above)
Comment (juuz0) [reply src/contentmanager.cpp:72]: Fixed in history

**Comment by juuz0** [comment]
Rebased fixups

**Pull request merged**